product-catalog/app/productcatalog/services/ProductProjectionService.java

```diff
@@ -0,0 +1,74 @@
+package productcatalog.services;
+
+import io.sphere.sdk.categories.Category;
+import io.sphere.sdk.client.PlayJavaSphereClient;
+import io.sphere.sdk.products.ProductProjection;
+import io.sphere.sdk.products.ProductVariant;
+import io.sphere.sdk.products.queries.ProductProjectionQuery;
+import io.sphere.sdk.products.search.ProductProjectionSearch;
+import io.sphere.sdk.queries.PagedQueryResult;
+import io.sphere.sdk.queries.PagedResult;
+import play.libs.F;
+
+import javax.inject.Inject;
+import java.util.*;
+
+public class ProductProjectionService {
+
+    private final PlayJavaSphereClient sphere;
+
+    @Inject
+    public ProductProjectionService(final PlayJavaSphereClient sphere) {
+        this.sphere = sphere;
+    }
+
+    public F.Promise<List<ProductProjection>> searchProducts(final int page, final int pageSize) {
+        final int offset = (page - 1) * pageSize;
+        return sphere.execute(ProductProjectionSearch.ofCurrent()
+                .withOffset(offset)
+                .withLimit(pageSize))
+                .map(PagedResult::getResults);
+    }
+
+    public F.Promise<Optional<ProductProjection>> searchProductBySlug(final Locale locale, final String slug) {
+        return sphere.execute(ProductProjectionQuery.ofCurrent().bySlug(locale, slug))
+                .map(PagedQueryResult::head);
+    }
+
+    public Optional<ProductVariant> findVariantBySku(final ProductProjection product, final String sku) {
+        return product.getAllVariants().stream()
+                .filter(variant -> variantHasSku(variant, sku))
+                .findFirst();
+    }
+
+    public F.Promise<List<ProductProjection>> getSuggestions(final List<Category> categories, final int numSuggestions) {
```

Comment:
For testing I would split this method into two.
One method has this signature and the other one has the same signatures plus a parameter of type Random and is called by the other one, so your can test the logic in a pure functional way since you can in tests pass a certain Random as parameter.